core-rust/state-manager/src/mempool/pending_transaction_result_cache.rs

```diff
@@ -14,46 +14,78 @@ pub type ExecutionRejectionReason = RejectionReason;
 #[derive(Debug, Clone, PartialEq, Eq)]
 pub enum MempoolRejectionReason {
     TransactionIntentAlreadyCommitted(AlreadyCommittedError),
+    SubintentAlreadyFinalized(SubintentAlreadyFinalizedError),
     FromExecution(Box<ExecutionRejectionReason>),
     ValidationError(TransactionValidationError),
 }
 
 #[derive(Debug, Clone, PartialEq, Eq)]
 pub struct AlreadyCommittedError {
-    pub notarized_transaction_hash: NotarizedTransactionHash,
     pub committed_state_version: StateVersion,
     pub committed_notarized_transaction_hash: NotarizedTransactionHash,
 }
 
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct SubintentAlreadyFinalizedError {
+    pub subintent_hash: SubintentHash,
+    pub committed_state_version: StateVersion,
+    pub committed_transaction_intent_hash: TransactionIntentHash,
+    pub committed_notarized_transaction_hash: NotarizedTransactionHash,
+}
+
 impl From<TransactionValidationError> for MempoolRejectionReason {
     fn from(value: TransactionValidationError) -> Self {
         Self::ValidationError(value)
     }
 }
 
 impl MempoolRejectionReason {
-    pub fn is_permanent_for_payload(&self) -> bool {
+    pub fn is_from_execution(&self) -> bool {
+        match self {
+            MempoolRejectionReason::TransactionIntentAlreadyCommitted(_) => false,
+            MempoolRejectionReason::SubintentAlreadyFinalized(_) => false,
+            MempoolRejectionReason::FromExecution(_) => true,
+            MempoolRejectionReason::ValidationError(_) => false,
```

Comment:
Per cuttlefish update, transaction validation is stateful. Is the permanent rejection logic below still correct?

Reply:
Not quite - if a Cuttlefish-only transaction is evaluated against Bottlenose it will be permanently rejected at validation time, but then become acceptable in Cuttlefish. 

I think this is unlikely to cause any critical issues -- and have some ideas for tweaking this in future, as discussed in the task list document -- but it's quite a hard subject to resolve satisfactorily without a lot of work. Happy to discuss later in the week.